### PR TITLE
update PartitionSpec with snapshot'schema

### PR DESCRIPTION
**Pull request by lurnagao-dahua**
Fix issues [11162](https://github.com/apache/iceberg/issues/11162)

### CODE REVIEW

**Comment by lurnagao-dahua** [comment]
Could you please take a review when you have time? @pvary @nastra @Fokko
I would greatly appreciate it!

**Comment by pvary** [comment]
Could we have a test for all of the documented use-cases:
```
-- time travel to October 26, 1986 at 01:21:00 -> uses the snapshot's schema
SELECT * FROM prod.db.table TIMESTAMP AS OF '1986-10-26 01:21:00';

-- time travel to snapshot with id 10963874102873L -> uses the snapshot's schema
SELECT * FROM prod.db.table VERSION AS OF 10963874102873;

-- time travel to the head of audit-branch -> uses the table's schema
SELECT * FROM prod.db.table VERSION AS OF 'audit-branch';
SELECT * FROM prod.db.table.`branch_audit-branch`;

-- time travel to the snapshot referenced by the tag historical-snapshot -> uses the snapshot's schema
SELECT * FROM prod.db.table VERSION AS OF 'historical-snapshot';
SELECT * FROM prod.db.table.`tag_historical-snapshot`;
```

File: core/src/test/java/org/apache/iceberg/DataTableScanTestBase.java [hunk 1]
```diff
@@ -309,4 +310,14 @@ public void testManifestLocationsInScanWithDeleteFiles() throws IOException {
             .get(0);
     assertThat(deletes.get(0).manifestLocation()).isEqualTo(deleteManifest.path());
   }
+
+  @TestTemplate
+  public void testTimeTravelScanWithAlterColumn() throws Exception {
```
Comment (pvary) [inline core/src/test/java/org/apache/iceberg/DataTableScanTestBase.java:315]: nit: no exception is thrown

File: core/src/main/java/org/apache/iceberg/TableMetadata.java [hunk 2]
```diff
@@ -719,7 +719,7 @@ public TableMetadata upgradeToFormatVersion(int newFormatVersion) {
     return new Builder(this).upgradeFormatVersion(newFormatVersion).build();
   }
 
-  private static PartitionSpec updateSpecSchema(Schema schema, PartitionSpec partitionSpec) {
+  public static PartitionSpec updateSpecSchema(Schema schema, PartitionSpec partitionSpec) {
```
Comment (pvary) [inline core/src/main/java/org/apache/iceberg/TableMetadata.java:722]: Could this be package private?

**Comment by nastra** [comment]
> Could we have a test for all of the documented use-cases:
> 
> ```
> -- time travel to October 26, 1986 at 01:21:00 -> uses the snapshot's schema
> SELECT * FROM prod.db.table TIMESTAMP AS OF '1986-10-26 01:21:00';
> 
> -- time travel to snapshot with id 10963874102873L -> uses the snapshot's schema
> SELECT * FROM prod.db.table VERSION AS OF 10963874102873;
> 
> -- time travel to the head of audit-branch -> uses the table's schema
> SELECT * FROM prod.db.table VERSION AS OF 'audit-branch';
> SELECT * FROM prod.db.table.`branch_audit-branch`;
> 
> -- time travel to the snapshot referenced by the tag historical-snapshot -> uses the snapshot's schema
> SELECT * FROM prod.db.table VERSION AS OF 'historical-snapshot';
> SELECT * FROM prod.db.table.`tag_historical-snapshot`;
> ```

I believe we should have that already in `TestSelect`. @lurnagao-dahua can you double-check please if there's a gap in `TestSelect` that doesn't cover any of these queries?

File: core/src/test/java/org/apache/iceberg/DataTableScanTestBase.java [hunk 3]
```diff
@@ -309,4 +310,14 @@ public void testManifestLocationsInScanWithDeleteFiles() throws IOException {
...
+    table.newFastAppend().appendFile(FILE_A).appendFile(FILE_B).commit();
+    long timeTravelSnapshotId = table.currentSnapshot().snapshotId();
+    table.updateSchema().renameColumn("id", "re_id").commit();
+    TableScan scan =
+        table.newScan().useSnapshot(timeTravelSnapshotId).filter(Expressions.equal("id", 5));
+    assertThat(Iterables.size(scan.planFiles())).isEqualTo(2);
```
Comment (nastra) [inline core/src/test/java/org/apache/iceberg/DataTableScanTestBase.java:321]:
```suggestion
    assertThat(scan.planFiles()).hasSize(2);
```

**Comment by lurnagao-dahua** [comment]
Thank you very much for the review. @pvary @nastra 
 I have made the corresponding changes. Could you please take a review when you have time?

File: core/src/main/java/org/apache/iceberg/BaseScan.java [hunk 4]
```diff
@@ -143,6 +147,14 @@ protected ExecutorService planExecutor() {
     return context().planExecutor();
   }
 
+  protected Map<Integer, PartitionSpec> specsById() {
+    List<PartitionSpec> specs =
+        Lists.newArrayList(
+            Iterables.transform(
+                table.specs().values(), spec -> TableMetadata.updateSpecSchema(schema, spec)));
```
Comment (pvary) [inline core/src/main/java/org/apache/iceberg/BaseScan.java:154]: What will happen if the spec is invalid for the given schema?
I see that the `TableMetadata` uses unchecked builds, but I was wondering if it is causing issues down the line.

Comment (lurnagao-dahua) [reply core/src/main/java/org/apache/iceberg/BaseScan.java]: Specifications verification is not required for replacement by referring to the [4361](https://github.com/apache/iceberg/pull/4361)
  I think validation is needed here, so I added it.

**Comment by pvary** [comment]
Based on the doc, the change looks good.
I would write an email to the dev list though, since this is a behavioural change, to highlight it to the community.

**Comment by mayankvadariya** [comment]
@lurnagao-dahua thank you for creating a PR, are there any plans to re-open this PR?

**Comment by lurnagao-dahua** [comment]
> @lurnagao-dahua thank you for creating a PR, are there any plans to re-open this PR?
I can reopen it at any time, but the community seems to have no plans.

**Comment by pvary** [comment]
@lurnagao-dahua: Could you please link the dev list discussion? Were there any comments on your message?

**Comment by mirageyjd** [comment]
Hi @lurnagao-dahua, are there any progress on this PR? I met the bug of inconsistent schema in time-travel query as well.

I have a different idea on how to fix this bug. The root cause is that evaluators such as `ProjectionEvaluator` and `InclusiveMetricsEvaluator` use schema in spec instead of schema defined in `TableScan`. Rather than reconstructing `PartitionSpec` with correct schema, it would be better to directly pass schema in `TableScan` to those evaluators. Reconstruction may fail since new `PartitionSpec` may not be compatible with old `Schema`. How do you think about it?

**Comment by lirui-apache** [comment]
It seems to me that the schema in PartitionSpec is ambiguous. IMO any schema that is compatible with the spec is fine, e.g. the schema when the spec is defined, or the current schema as long as it's still compatible. Would it be cleaner if we make this schema private to PartitionSpec? Evaluators needing a schema should get one in a more explicit way, e.g. get current schema from table or snapshot schema from the scan, depending on what they intend to do.

**Pull request closed without merging**